sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/InputFileBuilder.java

```diff
@@ -93,19 +98,52 @@ DefaultInputFile create(File file) {
   DefaultInputFile completeAndComputeMetadata(DefaultInputFile inputFile, InputFile.Type type) {
     inputFile.setType(type);
     inputFile.setModuleBaseDir(fs.baseDir().toPath());
-    inputFile.setCharset(fs.encoding());
+    Charset charset = detectCharset(inputFile.file(), fs.encoding());
+    inputFile.setCharset(charset);
```

Comment:
This is more a question for PO/support but if we start having files with different encoding, we should probably log in DEBUG what is the detected charset. @bellingard WDYT? To avoid too verbose logs, we could maybe log in a single line detected language + charset?